templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,871 @@
+
+    print(f"Number of checkpointed parameters: {get_number_of_params(model)}")
+
+    print("Model initialized with pretrained weights. Training starting...")
+    if not config["no_grad_ckpt"]:
+        model.gradient_checkpointing_enable()
+
+    optimizer_cls = (
+        torch.optim.AdamW
+        if accelerator.state.deepspeed_plugin is None
+        or "optimizer" not in accelerator.state.deepspeed_plugin.deepspeed_config
+        else DummyOptim
+    )
+
+    optimizer = optimizer_cls(
+        model.parameters(),
+        lr=lr,
+        betas=OPTIM_BETAS,
+        weight_decay=OPTIM_WEIGHT_DECAY,
+        eps=OPTIM_EPS,
+    )
+
+    # Instantiate scheduler
+    # Creates Dummy Scheduler if `scheduler` was specified in the config file or
+    # else, creates `args.lr_scheduler_type` Scheduler
+    # get train and valid dataset lengths
+
+    num_steps_per_epoch = math.ceil(train_ds_len / config["batch_size_per_device"])
+    total_training_steps = (
+        num_steps_per_epoch * num_epochs // gradient_accumulation_steps
+    )
+
+    if (
+        accelerator.state.deepspeed_plugin is None
+        or "scheduler" not in accelerator.state.deepspeed_plugin.deepspeed_config
+    ):
+        lr_scheduler = get_linear_schedule_with_warmup(
+            optimizer=optimizer,
+            num_warmup_steps=NUM_WARMUP_STEPS * num_devices,
+            num_training_steps=total_training_steps * num_devices,
+        )
+    else:
+        lr_scheduler = DummyScheduler(
+            optimizer,
+            warmup_num_steps=NUM_WARMUP_STEPS * num_devices,
+            total_num_steps=total_training_steps * num_devices,
+        )
+
+    # Prepare everything
+    # There is no specific order to remember, we just need to unpack the objects in the
+    # same order we gave them to the prepare method.
+    s = time.time()
+    model, optimizer, lr_scheduler = accelerator.prepare(model, optimizer, lr_scheduler)
+    print(f"Prepare done in {time.time() - s} seconds.")
+
+    # Now we train the model
+    if accelerator.is_main_process:
+        print("Starting training ...")
+        print("Number of batches on main process", train_ds_len // batch_size)
+
+    for epoch in range(num_epochs):
+        fwd_time_sum, bwd_time_sum, optim_step_time_sum = 0, 0, 0
+        s_epoch = time.time()
+        model.train()
+        loss_sum = torch.tensor(0.0).to(accelerator.device)
+
+        train_dataloader = train_ds.iter_torch_batches(
+            batch_size=batch_size,
+            collate_fn=collate_partial,
+        )
+
+        for step, batch in tqdm.tqdm(
+            enumerate(train_dataloader), total=train_ds_len // batch_size + 1
+        ):
+
+            # We could avoid this line since we set the accelerator with
+            # `device_placement=True`.
+            with accelerator.accumulate(model):
+                s_fwd = time.time()
+                outputs = model(**batch)
+                loss = outputs.loss
+                loss_sum += loss
```

Comment:
We had a bug fix the other day, where this line became `loss.item()`. Please make sure that the code in this PR is based off of the freshest OSS template commit!

Reply:
Fixed. Training code is now the same as the master